proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,664 @@
+
+```swift
+actor Worker {
+  func hi() {}
+}
+
+Task(on: actor) { actor in
+  actor.hi()
+}
+```
+
+The APIs added to Task are similar to the ones discussed above using executors as parameters, but specialized for Actor types:
+
+```swift
+extension Task where Failure == Never {
+  @discardableResult
+  init<TargetActor>(
+    on actor: TargetActor,
+    priority: TaskPriority? = nil,
+    operation: @Sendable @escaping (isolated TargetActor) async -> Success
+  ) where TargetActor: Actor
+
+  @discardableResult
+  static func detached<TargetActor>(
+    on actor: TargetActor,
+    priority: TaskPriority? = nil,
+    operation: @Sendable @escaping (isolated TargetActor) async -> Success
+  ) where TargetActor: Actor
+}
+
+extension Task where Failure == Error { 
+  @discardableResult
+  public init<TargetActor>(
+    on actor: TargetActor,
+    priority: TaskPriority? = nil,
+    operation: @Sendable @escaping (isolated TargetActor) async throws -> Success
+  ) where TargetActor: Actor
+  
+  @discardableResult 
+  static func detached<TargetActor>(
+    on actor: TargetActor,
+    priority: TaskPriority? = nil,
+    operation: @Sendable @escaping (isolated TargetActor) async throws -> Success
+  ) where TargetActor: Actor
+}
+```
+
+The same kind of APIs are offered for creating structured tasks using TaskGroups:
+
+```swift
+extension (Discarding)(Throwing)TaskGroup { 
+  func addTask<TargetActor: Actor>(
+    on: TargetActor, 
+    operation: (isolated TargetActor) async (throws) -> Success)
+}
+```
+
+#### Task executor preference and default actor isolated methods
+
+It is also worth explaining the interaction with actors which do not declare any executor requirement, like most actors.
+Such actors are referred to as "default actors" and are the default way of how actors are declared:
+
+```swift
+actor RunsAnywhere { // a "default" actor == without an executor requirement
+  func hello() {
+    return "Hello"
+  }
+}
+```
+
+Such actor has no requirement as to where it wants to execute. This means that if we were to call the `hello()` isolated 
+actor method from a task that has defined an executor preference -- the hello() method would still execute on a thread owned by that executor (!),
+however isolation is still guaranteed by the actor's semantics:
+
+```swift
+let anywhere = RunsAnywhere()
+Task { await anywhere.hello() } // runs on "default executor", using a thread from the global pool
+
+Task(on: myExecutor) { await anywhere.hello() } // runs on preferred executor, using a thread owned by that executor
```

Comment:
I’m not sure if I understand correctly, but doesn’t this part indicate isolated actor method jobs for an actor instance can be enqueued into multiple executors (default actor queue and preferred executor’s queue)? Doesn't it lead concurrent execution on an actor instance?

Reply:
In this version of the proposal the preferred executor takes precedence; We're not "running on" the actor's executor; though we'd have to use it for runtime assume-isolated checking.

Reply:
Ah, that makes sense 👍